### DISCUSSION TITLE
[ERROR] VImage::option()->set

### DISCUSSION THREAD

**Discussion by fakecan** [libvips/libvips#4273] [Q&A] [answered]
#### Describe the bug

```C++
VImage::option()->set
```

### To Reproduce

The code below works fine when the "VImage::option()->set" option is not used.

```C++
VImage thumbnail = VImage::thumbnail(
            imagePath.toStdString().c_str(),
            thumbnail_width,
            VImage::option()->set("height", thumbnail_height));
```

But using only the "VImage::option()->set" option will result in the following error:

```C++
HEAP[?.exe]:
Invalid address specified to RtlFreeHeap(?, ?)
```

Doesn't seem to be a value or type issue with the variable `imagePath.toStdString().c_str(), thumbnail_width, thumbnail_height`.

### Environment

 - OS : Windows 10
 - vips : 8.16

How do we fix it?
Thanks in advance.

**Comment by jcupitt** [comment]
Hi @fakecan,

It should work. I made you a tiny demo program:

```C++
/* compile with:
 *
 * g++ -g -Wall thumbnail.cpp `pkg-config vips-cpp --cflags --libs`
 */

#include <vips/vips8>

int
main(int argc, char **argv)
{   
    if (VIPS_INIT(argv[0]))
        return -1;

    vips::VImage thumbnail = vips::VImage::thumbnail(argv[1],
            200,
            vips::VImage::option()->set("height", 300));

    thumbnail.write_to_file(argv[2]);

    return 0;
}
```

I can run it with eg.:

```shell
$ ./a.out  ~/pics/k2.jpg x.jpg
```

And it generates a thumbnail.

If it's not working, you have probably not set your system up correctly. You need to supply a lot more information: how did you install libvips, what compiler are you using, what flags did you compile with, post a complete short program that reproduces the problem, etc.

**Comment by fakecan** [comment]
I was expecting a little bit of that. The compiler is expected to be captured. (Qt 6.8.0 in Qt Creator)
vips is being used in the CMakeLists.txt file using the downloaded windows binary file in the following way

```cmake
set(LIBVIPS_ROOT “C:/libvips”)
link_directories(“${LIBVIPS_ROOT}/lib”)
target_link_libraries(PROJECT_NAME PRIVATE
Qt${QT_VERSION_MAJOR}::Widgets
Qt${QT_VERSION_MAJOR}::Concurrent
vips-cpp
vips
glib-2.0
gobject-2.0
)
```

![Kits캡처](https://github.com/user-attachments/assets/0e61e933-4428-403a-81d1-ad01d4015583)
![image](https://github.com/user-attachments/assets/20603941-7921-4804-9d36-ce410e6abb91)


When you are working in Qt Creator and there is no “VOption::set” code, build&run is fine.
Any help would be appreciated. Thanks.

p.s. I'm sorry, but I think it's actually because I'm not familiar with the compiler part.

**Reply by jcupitt** [reply]
I would guess it's a C++ compiler version mismatch.

We compile the window libvips C++ binding with a recent clang, and you are using what looks like a mixture of g++ and MSVC. Since C++ does not define an ABI, you *MUST* use exactly the same compiler (and compiler version!) for every dependency in your project. 

I would copy the libvips C++ binding sources here:

https://github.com/libvips/libvips/tree/master/cplusplus

into your project and compile them with all your other code. It's simple to build (phew). 

You don't need to rebuild the whole of libvips (it has a C interface, which is standardised), just the small C++ layer.

**Reply by fakecan** [reply]
I think that probably solved it. I used Windows 10, MSYS2 settings.
I'll test a bit more, and if it works, I'll leave those settings as a reply.

libvips i love you, my goat

**Reply by fakecan** [reply]
Below is how I solved it.(1~2).
Question(3)

## 1. Environments
- OS : Windows 11
- IDE : Qt Creator 14.0.2 or 15.0.0
- Install libvips with MSYS2 MinGW64
![image](https://github.com/user-attachments/assets/00298e83-ce0a-4301-bca6-6e2afd8c967e)

## 2. CMakeLists.txt
```cmake
cmake_minimum_required(VERSION 3.16)

project(your_project_name)

set(CMAKE_AUTOUIC ON)
set(CMAKE_AUTOMOC ON)
set(CMAKE_AUTORCC ON)

set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

find_package(Qt6 REQUIRED COMPONENTS Widgets Concurrent)

find_package(PkgConfig REQUIRED)

pkg_check_modules(VIPS REQUIRED IMPORTED_TARGET vips)
pkg_check_modules(VIPS_CPP REQUIRED IMPORTED_TARGET vips-cpp)

qt_standard_project_setup()

qt_add_executable(your_project_name
    main.cpp
    mainwindow.cpp
    mainwindow.h
    mainwindow.ui
)

target_link_libraries(your_project_name PRIVATE
    Qt6::Widgets
    Qt6::Concurrent
    PkgConfig::VIPS
    PkgConfig::VIPS_CPP
)
```

## 3. Question
![스크린샷 2024-12-11 163824323](https://github.com/user-attachments/assets/994b9721-b144-4809-a406-52247af9fcec)

The msys2 package list looks like this, are the versions here persisted? Or are some of them deleted with each version upgrade? I think this is important for software shape management.(http://repo.msys2.org/)

**Reply by jcupitt** [reply, marked as answer]
Sorry, no idea, I'm not involved with the msys package.

If you want to make a windows program, you're probably better off with the DLLs we make and distribute:

https://github.com/libvips/build-win64-mxe/releases/tag/v8.16.0

**Reply by fakecan** [reply]
I recently created an environment using libvips and libtorch with MSVC on Windows 11 using the link you mentioned in the github issue.(I'll ask this as a separate discussion.)
The issue here is resolved, so I'll close it, thanks.
